controlplane/pkg/nsm/nsm_heal_processor.go

```diff
@@ -16,47 +16,141 @@ import (
 )
 
 type networkServiceHealProcessor interface {
-	healDstDown(healID string, connection *model.ClientConnection) bool
-	healDataplaneDown(healID string, connection *model.ClientConnection) bool
-	healDstUpdate(healID string, connection *model.ClientConnection) bool
-	healDstNmgrDown(healID string, connection *model.ClientConnection) bool
+	Heal(connection nsm.ClientConnection, healState nsm.HealState)
 }
 
-type nsmHealProcessor struct {
+type healProcessor struct {
 	serviceRegistry serviceregistry.ServiceRegistry
 	model           model.Model
 	properties      *nsm.NsmProperties
 
 	conManager connectionManager
 	nseManager networkServiceEndpointManager
+
+	eventCh chan healEvent
 }
 
 type connectionManager interface {
 	request(ctx context.Context, request networkservice.Request, existingConnection *model.ClientConnection) (connection.Connection, error)
-	close(ctx context.Context, clientConnection *model.ClientConnection, closeDataplane bool, modelRemove bool) error
+	Close(ctx context.Context, clientConnection nsm.ClientConnection) error
+}
+
+type healEvent struct {
+	healID    string
+	cc        *model.ClientConnection
+	healState nsm.HealState
 }
 
-func (p *nsmHealProcessor) healDstDown(healID string, connection *model.ClientConnection) bool {
+func newNetworkServiceHealProcessor(
+	serviceRegistry serviceregistry.ServiceRegistry,
+	model model.Model,
+	properties *nsm.NsmProperties,
+	conManager connectionManager,
+	nseManager networkServiceEndpointManager) networkServiceHealProcessor {
+
+	p := &healProcessor{
+		serviceRegistry: serviceRegistry,
+		model:           model,
+		properties:      properties,
+		conManager:      conManager,
+		nseManager:      nseManager,
+		eventCh:         make(chan healEvent, 1),
+	}
+	go p.serve()
+
+	return p
+}
+
+func (p *healProcessor) Heal(connection nsm.ClientConnection, healState nsm.HealState) {
+	healID := create_logid()
+	logrus.Infof("NSM_Heal(%v) %v", healID, connection)
+
+	cc := connection.(*model.ClientConnection)
+	if cc.ConnectionState != model.ClientConnectionReady {
+		//means that we already closing/healing
+		return
+	}
+
+	if !p.properties.HealEnabled {
+		logrus.Infof("NSM_Heal(%v) Is Disabled/Closing connection %v", healID, connection)
+
+		err := p.conManager.Close(context.Background(), cc)
+		if err != nil {
+			logrus.Errorf("NSM_Heal(%v) Error in Close: %v", healID, err)
+		}
+		return
+	}
+
+	p.model.ApplyClientConnectionChanges(cc.GetID(), func(cc *model.ClientConnection) {
+		cc.ConnectionState = model.ClientConnectionHealing
+	})
+
+	p.eventCh <- healEvent{
+		healID:    healID,
+		cc:        cc,
+		healState: healState,
+	}
+}
+
+func (p *healProcessor) serve() {
+	for {
+		e := <-p.eventCh
+		if e.cc == nil {
+			return
```

Comment:
Should we stop the healing goroutine? Maybe `continue` is more appropriate in that case?

Reply:
Changed to:
```
e, ok := <-p.eventCh
if !ok {
    return
}
```